### PR TITLE
fix: [DHIS2-8202] Implementing import for relationships

### PR DESCRIPTION
**Pull request by enricocolasante**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/TrackerIdentifierCollector.java [hunk 1]
```diff
@@ -134,6 +125,28 @@ private static void collectEvents(
         } );
     }
 
+    private static void collectRelationships(
+        Map<Class<?>, Set<String>> map, TrackerIdentifier identifier, List<Relationship> relationships )
+    {
+        relationships.forEach( relationship -> {
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/TrackerIdentifierCollector.java]: can this loop use `.parallelStream`?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java [hunk 2]
```diff
@@ -278,6 +294,49 @@ private TrackerTypeReport handleEvents( Session session, TrackerBundle bundle )
         return typeReport;
     }
 
+    private TrackerTypeReport handleRelationships( Session session, TrackerBundle bundle )
+    {
+        List<Relationship> relationships = bundle.getRelationships();
+        TrackerTypeReport typeReport = new TrackerTypeReport( TrackerType.RELATIONSHIP );
+
+        relationships.forEach( o -> bundleHooks.forEach( hook -> hook.preCreate( Relationship.class, o, bundle ) ) );
+        session.flush();
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java]: Explicit session flush is quite extreme? Can you add a comment why it is required?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            toRelationship.setTo( convertRelationShipType( fromRelationship.getTo() ) );
+            toRelationship.setLastUpdated( fromRelationship.getLastUpdated().toString() );
+            toRelationship.setRelationshipName( fromRelationship.getName() );
+            toRelationship.setRelationshipType( fromRelationship.getRelationshipType().getUid() );
+
+            return toRelationship;
+        } ).collect( Collectors.toList() );
+    }
+
+    private RelationshipItem convertRelationShipType( org.hisp.dhis.relationship.RelationshipItem from )
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java]: convertRelationShipType -> convertRelationshipType

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        {
+            return null;
+        }
+
+        return relationships.get( 0 );
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public List<Relationship> to( List<org.hisp.dhis.relationship.Relationship> relationships )
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java]: This class converts objects from one type to another, do we need to run the conversion in a database TX? Can't we cache the lookup data and run outside a tx?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java [hunk 5]
```diff
@@ -278,6 +294,49 @@ private TrackerTypeReport handleEvents( Session session, TrackerBundle bundle )
...
+            session.persist( toRelationship );
+            typeReport.getStats().incCreated();
+
+            if ( FlushMode.OBJECT == bundle.getFlushMode() )
+            {
+                session.flush();
+            }
+        }
+
+        session.flush();
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java]: Lots of flushing :) why can't we delegate the flushing to hibernate?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/TrackerIdentifierCollector.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/TrackerIdentifierCollector.java]: Done

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java (hunk 3 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java]: Done

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java (hunk 5 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java]: Removed the flushing

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java (hunk 4 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java]: Transactional here was completely wrong.
Removing it

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java (hunk 2 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java]: Removing the flushing but the one that is called when the user select the option to flush every time and object is added.

**Pull request merged**